Allow any_errors_fatal to be set in playbook - fixes #30691

s-hertel opened the pull request:
##### SUMMARY
With this any_errors_fatal may be set in the playbook as well as the config file (only the config file is currently noticed). Fixes #30691.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 lib/ansible/playbook/play_context.py
lib/ansible/plugins/strategy/linear.py

##### ANSIBLE VERSION
```
2.5.0
```

s-hertel committed:
Allow any_errors_fatal to be set in playbook.

bcoca commented on lib/ansible/plugins/strategy/linear.py:
```diff
@@ -252,7 +252,7 @@ def run(self, iterator, play_context):
                         run_once = templar.template(task.run_once) or action and getattr(action, 'BYPASS_HOST_LOOP', False)
 
-                        if (task.any_errors_fatal or run_once) and not task.ignore_errors:
+                        if (play_context.any_errors_fatal or run_once) and not task.ignore_errors:
```
this prevents the play setting from being overriden by block/role/tasks/etc.

s-hertel replied:
This seems to work now for blocks, roles, and tasks. @bcoca When you have time, can you look at this again?

bcoca replied:
from the changes this will now be ignored when set at role/block or task level, only play level is valid (rest are still settable, but ignored)

s-hertel commented:
Closing until I have time to fix this.

s-hertel committed:
Default to the config file value for any_errors_fatal only if it isn't already provided.

bcoca commented on lib/ansible/playbook/base.py:
```diff
@@ -165,7 +165,7 @@ class Base(with_metaclass(BaseMeta, object)):
     _check_mode          = FieldAttribute(isa='bool')
     _diff                = FieldAttribute(isa='bool')
-    _any_errors_fatal     = FieldAttribute(isa='bool', default=C.ANY_ERRORS_FATAL, always_post_validate=True)
+    _any_errors_fatal     = FieldAttribute(isa='bool', always_post_validate=True)
```
so this change i think is the base of the fix, but instead of what you are doing below, just make sure you override the `post validate` for the `play` object's `any_errors_fatal` and assign the default there if `None`. The rest should fall into place via inheritance.

bcoca replied:
The other option is to remove here but post validate at task level after checking inhertiance, set to the default if `None`, removing also `always_post_validate`

s-hertel committed:
Add post validate method in Play for any_errors_fatal.

Allow unrequired attributes to use post validate methods.

Template any_errors_fatal in post validate

add post validation to task

s-hertel committed:
remove post validation and add _get_attr method
